### DISCUSSION TITLE
Printing multiple selections

### DISCUSSION THREAD

**Discussion by hedbeerSICKAG** [PDConSec/vsc-print#351] [Ideas]
First, thanks for your work on this excellent extension!

I find the ability to print only the currently selected lines very useful, and think it would be nice if it would also work with multiple selections. (Pressing ctrl and selecting an additional range of lines is one way to get multiple selections. Pressing alt-enter when searching is another, useful especially in combination with regex searches.)

I found this closed issue #162 from 2022, but I hope it is OK I reopened the topic as a discussion.

Currently the extension seems to only print the first selection range. I saw the extension supports printing multiple selected files, so hopefully code for that can reused to print multiple selections. 

I had a quick look around, and think a starting point would be to use the `TextEditor` [property](https://code.visualstudio.com/api/references/vscode-api#TextEditor) `selections` instead of `selection` on [line 63 in print-session.ts](https://github.com/PDConSec/vsc-print/blob/6bf1615d673eb010a1517bb79f889f0987a800ff/src/print-session.ts#L63). I might try my hand on an MR later, but realistically, probably not. :(

**Comment by PeterWone** [comment]
Since you are the second person to tell me that multiple selections are possible I tried harder searching for this. Pressing `Ctrl` while making a second selection doesn't work for me, same as when I closed #162 - however holding down `Alt` does. 

- It will be some time before I can give this the time it requires.
- If you get there first, properly thought-out contributions are always well received so feel free to ask for assistance or explanation of parts of the code that don't make sense to you.
- Think about how it should be presented 
   - extra whitespace between the selections? vertical ellipses?
   - where an entire selection is heavily indented, should the indent be reduced to reduce wrapping? does this need to be a setting?

**Reply by hedbeerSICKAG** [reply]
I had a look, and made a sketch:

- In `print_session.ts`, block starting at [line 80](https://github.com/PDConSec/vsc-print/blob/6bf1615d673eb010a1517bb79f889f0987a800ff/src/print-session.ts#L80), loop over `editor!.selections` and turn variables `selectedText` and `startLine` into arrays.
- In `HtmlDocumentBuilder`, change members `code` and `startLine` from strings to arrays, adapt all code that set or use them.
- At [line 144](https://github.com/PDConSec/vsc-print/blob/6bf1615d673eb010a1517bb79f889f0987a800ff/src/renderers/html-document-builder.ts#L144), call `documentRenderer.getBodyHtml` for each set of text and startline, add HTML for divider, and concatenate into `bodyHtml`.

Does this sound like a reasonable approach? Or would you prefer adding a separate case like for the multi-file-selection?

Regarding presentation:
-  I think extra whitespace should be enough. With line numbers, a gap in the numbers column is clear. Without line numbers it is less good, but still reasonable. 
- Adapting the indentation does not feel like a necessary feature. Could be a nice setting, but not one which should be on by default.

**Reply by PeterWone** [reply]
Your sketch is exactly how I would have done it. Multi-file is more complicated because there is no guarantee all the files will have the same language in them eg a dotnet project web project is mostly C# but may include js script and powershell deployment scripts.

I agree about the indentation adjustment being a separate feature ticket. 

The same for separation. It ought to be a setting - {blank line|vertical ellipsis|horizontal rule} and this entails a new setting code to honour it and all the associated localisation for the 19 supported languages. That's a ticket in its own right. For now just embed whitespace, use `<div class="selection-leading">&nbsp;</div>` and define `div.selection-leading` in `src/css/settings.css` because I'll probably make the height a setting.
